fdbclient/SpecialKeySpace.actor.cpp

```diff
@@ -2717,6 +2722,279 @@ Future<Optional<std::string>> DataDistributionImpl::commit(ReadYourWritesTransac
+}
+
+void BulkLoadCancelImpl::clear(ReadYourWritesTransaction* ryw, const KeyRef& key) {
+	throw not_implemented();
+}
+
+void BulkLoadCancelImpl::set(ReadYourWritesTransaction* ryw, const KeyRef& key, const ValueRef& value) {
+	throw not_implemented();
+}
+
+ACTOR static Future<Optional<std::string>> BulkLoadingCancelCommitActor(ReadYourWritesTransaction* ryw,
+                                                                        KeyRangeRef kr) {
+	// Check if there are existing bulk loading new task in the current transaction
+	if (existingBulkLoadUpdate(ryw, Standalone(KeyRangeRef("task/"_sr, "task0"_sr)))) {
+		TraceEvent(SevWarn, "BulkLoadTaskCommitError")
+		    .detail("Reason", "Exist bulk loading cancel in the same transaction");
+		throw bulkload_cancel_task_input_error();
+	}
+
+	// Get all cancelled range in the current transaction
+	state KeyRange cancelRange =
+	    Standalone(KeyRangeRef("cancel/"_sr, "cancel0"_sr))
+	        .withPrefix(SpecialKeySpace::getModuleRange(SpecialKeySpace::MODULE::BULKLOADING).begin);
+	state std::vector<KeyRange> cancelRanges;
+	auto ranges = ryw->getSpecialKeySpaceWriteMap().containedRanges(cancelRange);
+	for (auto iter = ranges.begin(); iter != ranges.end(); ++iter) {
+		if (!iter->value().first) {
+			continue;
+		}
+		ASSERT(!iter->value().second.present());
+		cancelRanges.push_back(iter->range().removePrefix(cancelRange.begin));
+	}
+	cancelRanges = coalesceRangeList(cancelRanges);
+
+	// Get all ranges to cancel intersecting the global task map
+	state int i = 0;
+	state Key readBeginKey;
+	state Key readEndKey;
+	state std::vector<KeyRange> rangesToRemove;
+	for (; i < cancelRanges.size(); i++) {
+		readBeginKey = cancelRanges[i].begin;
+		readEndKey = cancelRanges[i].end;
+		while (readBeginKey < readEndKey) {
+			KeyRange rangeToRead = Standalone(KeyRangeRef(readBeginKey, readEndKey));
+			RangeResult result = wait(krmGetRanges(&(ryw->getTransaction()), bulkLoadPrefix, rangeToRead));
+			for (int j = 0; j < result.size() - 1; j++) {
+				if (!result[j].value.empty()) {
+					KeyRange existRange = Standalone(KeyRangeRef(result[j].key, result[j + 1].key));
+					BulkLoadState existBulkLoadTask = decodeBulkLoadState(result[j].value);
+					ASSERT(existBulkLoadTask.isValid());
+					rangesToRemove.push_back(existBulkLoadTask.range);
+				}
+			}
+			readBeginKey = result.back().key;
+		}
+	}
+	rangesToRemove = coalesceRangeList(rangesToRemove);
+
+	// Remove task from the global task map
+	i = 0;
+	for (; i < rangesToRemove.size(); i++) {
+		wait(krmSetRange(&(ryw->getTransaction()), bulkLoadPrefix, rangesToRemove[i], Value()));
+		TraceEvent("BulkLoadCancelCommitEach")
+		    .detail("Range", rangesToRemove[i].toString())
+		    .detail("KR", kr.toString());
+	}
+
+	return Optional<std::string>();
+}
+
+Future<Optional<std::string>> BulkLoadCancelImpl::commit(ReadYourWritesTransaction* ryw) {
+	return BulkLoadingCancelCommitActor(ryw, getKeyRange());
+}
+
```

Comment:
Is it possible to write unit tests for this range 'math'; i.e. checking inside range with prefix, w/o prefix etc.